iothub/device/src/Transport/AmqpIoT/AmqpIoTSession.cs

```diff
@@ -218,11 +222,33 @@ TimeSpan timeout
                 amqpLinkSettings.AddProperty(AmqpIoTErrorAdapter.ChannelCorrelationId, CorrelationId);
             }
 
-            SendingAmqpLink sendingLink = new SendingAmqpLink(amqpLinkSettings);
-            sendingLink.AttachTo(amqpSession);
-            await sendingLink.OpenAsync(timeout).ConfigureAwait(false);
-            if (Logging.IsEnabled) Logging.Exit(typeof(AmqpIoTSession), deviceIdentity, $"{nameof(OpenSendingAmqpLinkAsync)}");
-            return new AmqpIoTSendingLink(sendingLink);
+            try
+            {
+                SendingAmqpLink sendingLink = new SendingAmqpLink(amqpLinkSettings);
+                sendingLink.AttachTo(amqpSession);
+                await sendingLink.OpenAsync(timeout).ConfigureAwait(false);
+                return new AmqpIoTSendingLink(sendingLink);
+            }
+            catch (Exception e) when (!e.IsFatal())
+            {
+                Exception ex = AmqpIoTExceptionAdapter.ConvertToIoTHubException(e, amqpSession);
+                if (ReferenceEquals(e, ex))
```

Comment:
Same as above - this should be encapsulated. (There are around 10 places where this code was pasted.)

Reply:
Can you propose ? Table

Reply:
@prmathur-microsoft see above: code should be encapsulated into a class/method/etc instead of copy/pasted to simplify maintenance and reduce places where we would need to fix bugs in the future.